src/main/java/com/google/devtools/build/lib/packages/RuleClass.java

```diff
@@ -941,7 +942,7 @@ public RuleClass build(String name, String key) {
             attr(STARLARK_BUILD_SETTING_DEFAULT_ATTR_NAME, type)
                 .nonconfigurable(BUILD_SETTING_DEFAULT_NONCONFIGURABLE)
                 .mandatory();
-        if (BuildType.isLabelType(type)) {
+        if (type.getLabelClass() == LabelClass.DEPENDENCY) {
```

Comment:
When is this true? It doesn't seem to be true at any call site and AFAIU in order to fix the associated bug, the `build_setting_default` attribute must not create any dependencies, and `allowedFilesTypes()` / `allowedRuleClasses()` only make sense if a dependency is created. Assert that it's not true instead? (either here or in `BuildSetting`)

Reply:
I changed this from `isLabelType()` to `== LabelClass.DEPENDENCY` specifically because of the change to `NODEP_LABEL`: with this change, the calls to `allowedFileTypes()` and `allowedRuleClasses()` were failing because they only accept real labels, not nodep labels.

Sorry I didn't wait for your review, this is currently internal at cl/369288702.

Reply:
Let me rephrase this:  when I look at the code, `RuleClass.getBuildSetting()` never returns a `Type` where `getLabelClass()` would return `DEPENDENCY`. But then why is this branch here?

In addition, if there was any case where that did return `DEPENDENCY`, that would trigger the bug you are fixing here, so why not assert the opposite?

Reply:
Ah, now I see your point. This entire check should be removed, since it's not possible to trigger it.

Should I add a check in `BuildSetting.create()` to ensure that only `NODEP_LABEL` and `NODEP_LABEL_CLASS` are used? I will make changes in the internal version and test this.

Reply:
I think a better approach is to assert `buildSetting.getType().getLabelClass() != LabelClass.DEPENDENCY` beacuse that says exactly what is needed: that the type shouldn't be one that creates a dependeny.

Reply:
But `allowedFileTypes` can only validly be called when the type is `LabelClass.DEPENDENCY` (see [the definitions](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/packages/Attribute.java;drc=d743e575c29ed954b1a2edbe5a6da66ccb29af8e;l=866)). Since it is now not possible to have an attribute with that type, there's no valid way to call this.